ubyssey/views.py

```diff
@@ -442,6 +448,109 @@ def newsletter(self, request):
+        else:
+            until_month = today.month + (months if months is not None else 12)
+            until_year = today.year
+            while until_month > 12:
+                until_month -= 12
+                until_year += 1
+            dt_until = today.replace(year=until_year, month=until_month)
+            events = events.filter(end_time__lte=dt_until)
+
+        if category is not None and category != 'all':
+            events = events.filter(category__exact=category)
+
+        HARD_MAX = 100
+        events = events[:HARD_MAX]
+
+        return events
+
+    def organize_events_by_date(self, events):
+        events_by_date = OrderedDict()
+
+        for event in events:
+            start = event.start_time # Zulu
+            start = start.astimezone(timezone('America/Vancouver'))
+            year = start.year
+            month = start.month
+            month = calendar.month_name[month]
+            day = '%s. %d' % (start.strftime('%a'), start.day)
+            if year not in events_by_date:
+                events_by_date[year] = OrderedDict()
+            if month not in events_by_date[year]:
+                events_by_date[year][month] = OrderedDict()
+            if day not in events_by_date[year][month]:
+                events_by_date[year][month][day] = []
+
+            events_by_date[year][month][day].append(event)
+
+        return events_by_date
+
+    def events_calendar(self, request):
+        category = request.GET.get('category')
+        months = request.GET.get('months')
+        start = request.GET.get('start')
+        end = request.GET.get('end')
+
+        events = self.get_calendar_events(category=category, months=months, start=start, end=end)
+        events_by_date = self.organize_events_by_date(events)
+
+        context = {
+            'meta': {
+                'title': 'Calendar',
+                'description': 'Ubyssey calendar of events',
+                'url': "%s%s" % (settings.BASE_URL, reverse('calendar'))
+            },
+            'events_by_date': events_by_date,
+            'this_year': date.today().year,
+            'category': category
+        }
+
+        return render(request, 'events/calendar.html', context)
+
+    def get_event_meta(self, event):
+        meta_image = None
+        if event.image:
+            meta_image = event.image.url
+
+        return {
+            'title': event.title,
+            'description': event.description,
+            'image': "%s%s" %(settings.BASE_URL, meta_image),
+            'url': "%sevent/%s/" % (settings.BASE_URL, event.id)
+        }
+
+    def event_detail(self, request, event_id):
+        # TODO: find a way to select without putting event_id in url
+        # eg. slug generated from date & name to avoid collision
+        event = get_object_or_404(Event, pk=event_id)
+
+        if event.is_submission or not event.is_published:
+            raise Http404('Event could not be found.')
```

Comment:
I think it'd prefer to perform the lookup all in one step:

```python
try:
  event = EventsHelpers.get_event(pk)
except:
  raise Http404('Event could not be found.'
```

Then in the helper:

```python
return Event.objects.get(pk=pk, is_submission=False, is_published=True)
```